src/zkevm_specs/state.py

```diff
@@ -510,11 +469,11 @@ def keys_rwc_to_limbs_in_order(row: Row) -> List[FQ]:
     elif row.tag() == Tag.Stack:
         check_stack(row, row_prev)
     elif row.tag() == Tag.Storage:
-        check_storage(row, row_prev, tables)
+        check_storage(row, row_prev, row_next, tables)
```

Comment:
I think a set of constraints is missing to forbid the prover from changing the root in unexpected rows.
Now we're only doing MPT lookup on the last entry for a unique key when field is Account or Storage.  So that's the only row where a root value change can happen.  I think we should constraint the rest of the rows to have a root which is the same as the previous row.

Otherwise, this example would pass the constraint but should be invalid:

| tag | address | root |
| ---  | ---           | ---   |
| TxRefund | A | X |
| Account | B | Y |
| Account | B | Z | 
| Memory | 1 | Z |

Here the root goes from X to Z, but only Y-Z transitions are MPT-proved, so in the transition X-Y the MPT could have an arbitrary change.

So I think the following constraint should be added:
```python
if not(not all_keys_eq(row, row_next) and (tag == Account or tag == Storage)):
    row.root == row_prev.root

Reply:
Haven't checked all the context yet. It looks like we can simplify
```
not(not all_keys_eq(row, row_next) and (tag == Account or tag == Storage))
```
to 

```
not(not all_keys_eq(row, row_next) and (tag in (Account, Storage))
```

to

```
all_keys_eq(row, row_next) or (tag not in (Account, Storage))
```

Reply:
I ended up adding `assert row.root == row_prev.root` for each of the tags, because the degree of the expression for `all_keys_eq(row, row_next) or (tag not in (Account, Storage))` would raise the total degree of the constraint system.

Reply:
Makes sense, addressing this :)